### DISCUSSION TITLE
libmesh compilation error: No rule to make target exodusII.h?

### DISCUSSION THREAD

**Discussion by alikhamze** [idaholab/moose#19956] [Q&A General] [answered]
Hi all,

I'm following the offline build guide [here](https://mooseframework.inl.gov/getting_started/installation/offline_installation.html) for installing Moose on a system without internet access. 
I was able to successfully complete the steps through the compilation of petsc, but have encountered an error when building libmesh.

Here is the error message that interrupts `update_and_rebuild_libmesh.sh`:
```
make[3]: *** No rule to make target '../../../contrib/exodusii/v5.22/exodus/cbind/include/exodusII.h', needed by `exodusII.h'.  Stop.
make[3]: Leaving directory '/home/sr2/ali.hamze/offline/moose/libmesh/build/include/libmesh'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/home/sr2/ali.hamze/offline/moose/libmesh/build/include'
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/sr2/ali.hamze/offline/moose/libmesh/build/include'
make: *** [all-recursive] Error 1
```

When I looked in `libmesh/contrib/exodusii/`, I saw there was actually no `v5.22` directory, but rather a `v5.22b` directory. I tried renaming `v5.22b` to `v5.22` and rerunning the `update_and_rebuild_libmesh.sh` script but encountered the same error.

The only other detail of the compilation process that I followed is that I was installing everything in place.

Has anyone encountered and fixed this error before?

Thanks!

**Comment by loganharbour** [comment, marked as answer]
Ping @roystgnr

**Reply by roystgnr** [reply]
This is fascinating.  I've never seen it before.

Could I see the output of `grep -i exodus /home/sr2/ali.hamze/offline/moose/libmesh/build/config.log`?  (I think that should be the right location; grep whereever your build dir config.log file is if I got it wrong.

**Reply by alikhamze** [reply]
Hi @roystgnr , here's what is returned by grep:
```
configure:55096: result: defining -DNOT_NETCDF4 for our Exodus build
configure:55135: result: <<< Configuring library with Exodus version v5.22 support >>>
configure:55152: result: <<< Configuring library with Exodus Fortran API >>>
| #define HAVE_EXODUS_API 1
| #define DETECTED_EXODUS_VERSION_MAJOR 5
| #define DETECTED_EXODUS_VERSION_MINOR 22
EXODUS_FORTRAN_API_FALSE='#'
EXODUS_FORTRAN_API_TRUE=''
EXODUS_INCLUDE='-I$(top_srcdir)/contrib/exodusii/v5.22/exodus/cbind/include'
EXODUS_NOT_NETCDF4_FLAG='-DNOT_NETCDF4'
LIBMESH_ENABLE_EXODUS_FALSE='#'
LIBMESH_ENABLE_EXODUS_TRUE=''
LIBMESH_ENABLE_EXODUS_V509_FALSE=''
LIBMESH_ENABLE_EXODUS_V509_TRUE='#'
LIBMESH_ENABLE_EXODUS_V522_FALSE='#'
LIBMESH_ENABLE_EXODUS_V522_TRUE=''
LIBMESH_ENABLE_EXODUS_V811_FALSE=''
LIBMESH_ENABLE_EXODUS_V811_TRUE='#'
libmesh_contrib_INCLUDES='-I$(top_srcdir)/contrib/metaphysicl/src/numerics/include -I$(top_srcdir)/contrib/metaphysicl/src/core/include -I$(top_srcdir)/contrib/metaphysicl/src/utilities/include -I$(top_srcdir)/contrib/nanoflann/include -I$(top_srcdir)/contrib/fparser -I$(top_srcdir)/contrib/libHilbert/include -I$(top_srcdir)/contrib/nemesis/v5.22/nemesis -I$(top_srcdir)/contrib/exodusii/v5.22/exodus/cbind/include -I$(top_srcdir)/contrib/netcdf/v4/include -I$(top_srcdir)/contrib/gmv -I$(top_srcdir)/contrib/qhull/qhull/src -I$(top_srcdir)/contrib/qhull/qhull/src/libqhullcpp -I$(top_builddir)/contrib/poly2tri/modified   -I$(top_srcdir)/contrib/tecplot/tecio/include -I$(top_srcdir)/contrib/gzstream -I$(top_srcdir)/contrib/boost/include -I$(top_builddir)/contrib/timpi/src/utilities/include -I$(top_srcdir)/contrib/timpi/src/utilities/include -I$(top_srcdir)/contrib/timpi/src/parallel/include -I$(top_srcdir)/contrib/timpi/src/algorithms/include '
#define HAVE_EXODUS_API 1
#define DETECTED_EXODUS_VERSION_MAJOR 5
#define DETECTED_EXODUS_VERSION_MINOR 22
```

**Reply by roystgnr** [reply]
Well, that's enough to steer me away from my only suspicion (that the new Exodus 8.11 configuration code was somehow fouling up something).

And looking at this more closely ... it's simply not finding the exodusII.h file for v5.22, because it's missing the v5.22 symbolic link to v5.22b?  That's baffling.  The symlink is part of the libMesh git repository, not even in a submodule.  If I clone libMesh, I see the symlink.  If I `rm contrib/exodusii/v5.22`, then `git status` reports the symlink as `deleted` in an unstaged change.  Are you on an OS or filesystem that doesn't support symbolic links?  Does `find /home/sr2/ali.hamze/offline/moose/libmesh/ -type l` show you any other symlinks? (either the ones in your build/ dir, or the other dozen-odd symlinks we have in contrib/ and our README, or both?)

**Reply by alikhamze** [reply]
Hi @roystgnr , I did find other symlinks with that command, so I don't think that was the issue.

I retried the compilation procedure because I realized that I hadn't set environment variables for the compilers for the libmesh step, and it seems to have worked! 
I'm not sure why the errors were so unclear--the only other thing I tried was tarring the inputs before transferring them rather than using rsync like I did before--but things seem to be working fine now.

I'll mark this as the answer, thanks for your time and your help!
